orbit/cmd/orbit/orbit.go

```diff
@@ -481,64 +499,111 @@ func main() {
 		fleetURL:        fleetURL,
 		deviceAuthToken: deviceAuthToken,
 		insecure:        insecure,
-		done:            make(chan struct{}),
+		interruptCh:     make(chan struct{}),
+		executeDoneCh:   make(chan struct{}),
 	}
 }
 
 func (d *desktopRunner) actor() (func() error, func(error)) {
 	return d.execute, d.interrupt
 }
 
+// execute makes sure the fleet-desktop application is running.
+//
+// We have to support the scenario where the user closes its sessions (log out).
+// To support this, we add retries to execuser.Run. Basically retry execuser.Run until it succeds,
+// which will happen when the user logs in.
+// Once fleet-desktop is started, the process is monitored (user processes get killed when the user
+// closes all its sessions).
+//
+// NOTE(lucas): This logic could be improved to detect if there's a valid session or not first.
 func (d *desktopRunner) execute() error {
-	log.Info().Str("path", d.appPath).Msg("opening")
+	defer close(d.executeDoneCh)
+
+	log.Info().Str("path", d.desktopPath).Msg("opening")
 	url, err := url.Parse(d.fleetURL)
 	if err != nil {
 		return fmt.Errorf("invalid fleet-url: %w", err)
 	}
 	url.Path = path.Join(url.Path, "device", d.deviceAuthToken)
-	opts := []open.AppOption{
-		open.AppWithEnv("FLEET_DESKTOP_DEVICE_URL", url.String()),
-		open.AppWithEnv("FLEET_DESKTOP_DEVICE_API_TEST_PATH", path.Join("api", "latest", "fleet", "device", d.deviceAuthToken)),
+	opts := []execuser.Option{
+		execuser.WithEnv("FLEET_DESKTOP_DEVICE_URL", url.String()),
+		execuser.WithEnv("FLEET_DESKTOP_DEVICE_API_TEST_PATH", path.Join("api", "latest", "fleet", "device", d.deviceAuthToken)),
 	}
 	if d.insecure {
-		opts = append(opts, open.AppWithEnv("FLEET_DESKTOP_INSECURE", "1"))
+		opts = append(opts, execuser.WithEnv("FLEET_DESKTOP_INSECURE", "1"))
 	}
-	if err := open.App(d.appPath, opts...); err != nil {
-		return fmt.Errorf("open desktop app: %w", err)
+
+	for {
+		// First retry logic to start fleet-desktop.
+		if done := retry(30*time.Second, d.interruptCh, func() bool {
+			// Orbit runs as root user on Unix and as SYSTEM (Windows Service) user on Windows.
+			// To be able to run the desktop application (mostly to register the icon in the system tray)
+			// we need to run the application as the login user.
+			// Package execuser provides multi-platform support for this.
+			if err := execuser.Run(d.desktopPath, opts...); err != nil {
+				log.Debug().Err(err).Msg("execuser.Run")
+				return true
+			}
+			return false
+		}); done {
+			return nil
+		}
+
+		// Second retry logic to monitor fleet-desktop.
+		if done := retry(30*time.Second, d.interruptCh, func() bool {
+			switch _, err := getProcessByName(constant.DesktopAppExecName); {
+			case err == nil:
+				return true // all good, process is running, retry.
+			case errors.Is(err, errProcessNotFound):
+				log.Debug().Msgf("%s process not running", constant.DesktopAppExecName)
+				return false // process is not running, do not retry.
```

Comment:
I don't understand, wouldn't this be one of the cases where you want to run fleet desktop?

Reply:
Oh, you return false here, so that we loop and start again... otherwise `retry` will keep retrying. Am I understanding this correctly?

Reply:
Correct:
- if the process is running then I want to continue the inner loop (continue calling getProcessByName).
- if the process is not running then I want to exit the inner loop and start running the `execuser.Run` loop above.